File: src/parser/pattern_cover.ml
(**
 * Copyright (c) Facebook, Inc. and its affiliates.
 *
 * This source code is licensed under the MIT license found in the
 * LICENSE file in the root directory of this source tree.
 *)

open Flow_ast
open Parser_common
open Parser_env

module type COVER = sig
  val as_expression : env -> pattern_cover -> (Loc.t, Loc.t) Expression.t

  val as_pattern : ?err:Parse_error.t -> env -> pattern_cover -> (Loc.t, Loc.t) Pattern.t

  val empty_errors : pattern_errors

  val rev_append_errors : pattern_errors -> pattern_errors -> pattern_errors

  val rev_errors : pattern_errors -> pattern_errors
end

module Cover (Parse : PARSER) : COVER = struct
  let as_expression env = function
    | Cover_expr expr -> expr
    | Cover_patt (expr, { if_expr; if_patt = _ }) ->
      List.iter (error_at env) if_expr;
      expr

  let as_pattern ?(err = Parse_error.InvalidLHSInAssignment) env cover =
    let expr =
      match cover with
      | Cover_expr expr -> expr
      | Cover_patt (expr, { if_expr = _; if_patt }) ->
        List.iter (error_at env) if_patt;
        expr
    in
    if not (Parse.is_assignable_lhs expr) then error_at env (fst expr, err);

    (match expr with
    | (loc, Flow_ast.Expression.Identifier (_, { Flow_ast.Identifier.name; comments = _ }))
      when is_restricted name ->
      strict_error_at env (loc, Parse_error.StrictLHSAssignment)
    | _ -> ());

    Parse.pattern_from_expr env expr

  let empty_errors = { if_patt = []; if_expr = [] }

  let rev_append_errors a b =
    { if_patt = List.rev_append a.if_patt b.if_patt; if_expr = List.rev_append a.if_expr b.if_expr }

  let rev_errors a = { if_patt = List.rev a.if_patt; if_expr = List.rev a.if_expr }
end

File: rust_port/crates/flow_parser/src/pattern_cover.rs
/*
 * Copyright (c) Meta Platforms, Inc. and affiliates.
 *
 * This source code is licensed under the MIT license found in the
 * LICENSE file in the root directory of this source tree.
 */

use std::ops::Deref;

use dupe::Dupe;

use crate::ast::expression::ExpressionInner;
use crate::ast::*;
use crate::expression_parser;
use crate::loc::Loc;
use crate::parse_error::ParseError;
use crate::parser_env::try_parse::Rollback;
use crate::parser_env::*;
use crate::pattern_parser;

#[derive(Debug, Clone)]
pub(super) enum PatternCover {
    CoverExpr(expression::Expression<Loc, Loc>),
    CoverPatt(expression::Expression<Loc, Loc>, PatternCoverErrors),
}

#[derive(Debug, Clone)]
pub(super) struct PatternCoverErrors {
    pub(super) if_expr: Vec<(Loc, ParseError)>,
    pub(super) if_patt: Vec<(Loc, ParseError)>,
}

const EMPTY_ERRORS: PatternCoverErrors = PatternCoverErrors {
    if_expr: Vec::new(),
    if_patt: Vec::new(),
};

impl PatternCoverErrors {
    pub(super) fn empty() -> Self {
        EMPTY_ERRORS
    }

    pub(super) fn append(&mut self, other: Self) {
        let Self {
            if_expr: mut other_if_expr,
            if_patt: mut other_if_patt,
        } = other;
        self.if_expr.append(&mut other_if_expr);
        self.if_patt.append(&mut other_if_patt);
    }
}

pub(super) fn as_expression(
    env: &mut ParserEnv,
    cover: PatternCover,
) -> Result<expression::Expression<Loc, Loc>, Rollback> {
    match cover {
        PatternCover::CoverExpr(expression) => Ok(expression),
        PatternCover::CoverPatt(expression, pattern_errors) => {
            for (loc, error) in pattern_errors.if_expr {
                env.error_at(loc, error)?;
            }
            Ok(expression)
        }
    }
}

pub(super) fn as_pattern(
    env: &mut ParserEnv,
    error: Option<ParseError>,
    cover: PatternCover,
) -> Result<pattern::Pattern<Loc, Loc>, Rollback> {
    let expr = match cover {
        PatternCover::CoverExpr(expression) => expression,
        PatternCover::CoverPatt(expression, pattern_errors) => {
            for (loc, error) in pattern_errors.if_patt {
                env.error_at(loc, error)?;
            }
            expression
        }
    };
    if !expression_parser::is_assignable_lhs(&expr) {
        env.error_at(
            expr.loc().dupe(),
            error.unwrap_or(ParseError::InvalidLHSInAssignment),
        )?;
    }

    match expr.deref() {
        ExpressionInner::Identifier { loc, inner } if is_reserved(&inner.name) => {
            env.strict_error_at((loc.dupe(), ParseError::StrictLHSAssignment))?
        }
        _ => {}
    }

    pattern_parser::from_expr(env, expr)
}
